### PR TITLE
feat: sap provider - support orchestration mode

### PR DESCRIPTION
**Pull request by yuvalman**
### Related Issue

**Issue:** #XXXX

can be merged only after #5315 is merged

### Description

- Add support for orchestration mode for users whose AI Core service instance plan is `extended`, **which provides all supported models in Cline without the need to manage deployments, as is required in Native API mode**. Orchestration mode is enabled by default.
- For other plans, maintain backward compatibility by continuing to use the deployed models in their AI Core service instance.
- Automatically detect whether orchestration mode is available in the AI Core service instance.

more details about the reason to create this pr can be found in this comment: https://github.com/cline/cline/issues/5058#issuecomment-3135402032

### Test Procedure

- checked edge cases such as changing ai core credentials
- check the ui changes, and how the backend logic. For example, when orchestration mode is checked, the backend uses the orchestration

### Type of Change

-   [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
-   [x] ✨ New feature (non-breaking change which adds functionality)
-   [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] ♻️ Refactor Changes
-   [ ] 💅 Cosmetic Changes
-   [ ] 📚 Documentation update
-   [ ] 🏃 Workflow Changes

### Pre-flight Checklist

-   [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
-   [x] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`)
-   [x] I have created a changeset using `npm run changeset` (required for user-facing changes)
-   [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Screenshots

before (no orchestration mode checkbox):
image

after:

for `extended` ai core service plan when orchestration mode box is checked (show one list with all the models supported by cline)

Screenshot 2025-08-13 at 17 07 00

for `extended` ai core service plan when orchestration mode box is **not** checked (show the seperated list of deploy and not deployed models):
Screenshot 2025-08-13 at 17 09 55

for other plans (such as `standard` and `free`), keep the current behavior (without the "Orchestration Mode" checkbox) that shows only the Model list (separated to deployed and not deployed):

Screenshot 2025-08-13 at 17 12 48

### Additional Notes

----

> [!IMPORTANT]
> Adds orchestration mode for SAP AI Core, enabling access to all models for 'extended' plan users, with UI updates and tests.
> 
>   - **Behavior**:
>     - Adds orchestration mode for SAP AI Core in `SapAiCoreProvider.tsx`, enabled by default for 'extended' plan users.
>     - Maintains backward compatibility for other plans by using deployed models.
>     - Automatically detects orchestration mode availability.
>   - **UI Components**:
>     - Adds `SapAiCoreModelPicker.tsx` for model selection, supporting orchestration and non-orchestration modes.
>     - Updates `SapAiCoreProvider.tsx` to include orchestration mode checkbox and model picker.
>   - **Testing**:
>     - Adds tests in `SapAiCoreModelPicker.spec.tsx` to verify model picker behavior in different modes.
>     - Mocks API and context in `setupTests.ts` for testing purposes.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for dfb5ee8deb70cf9c22a631ba1ec85d2a32cab5ba. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: docs/provider-config/sap-aicore.mdx [hunk 1]
```diff
@@ -32,8 +32,26 @@ Refer to the [Generative AI Hub Supported Models page](https://me.sap.com/notes/
 5.  **Enter Base URL:** Add the `.serviceurls.AI_API_URL` field from the service binding into the "AI Core Base URL" field.
 6.  **Enter Auth URL:** Add the `.url` field from the service binding into the "AI Core Auth URL" field.
 7.  **Enter Resource Group:** Add the resource group where you have your model deployments. See [Create a Deployment for a Generative AI Model](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/create-deployment-for-generative-ai-model-in-sap-ai-core).
-8.  **Select Model:** Choose your desired model from the "Model" dropdown.
+8.  **Configure Orchestration Mode:** If you have an `extended` service plan, the "Orchestration Mode" checkbox will be automatically appeared.
```
Comment (ncryptedV1) [inline docs/provider-config/sap-aicore.mdx]:
```suggestion
8.  **Configure Orchestration Mode:** If you have an `extended` service plan, the "Orchestration Mode" checkbox will automatically appear.
```

File: docs/provider-config/sap-aicore.mdx [hunk 2]
```diff
@@ -32,8 +32,26 @@ Refer to the [Generative AI Hub Supported Models page](https://me.sap.com/notes/
...
+9.  **Select Model:** Choose your desired model from the "Model" dropdown.
+
+### Orchestration Mode vs Native API
+
+**Orchestration Mode:**
+- **Requires Extended Plan:** Only available with SAP AI Core `extended` service plan
```
Comment (ncryptedV1) [inline docs/provider-config/sap-aicore.mdx]:
```suggestion
- **Requires Generative AI Hub:** Only available with SAP AI Core `extended` service plan
```

File: docs/provider-config/sap-aicore.mdx [hunk 3]
```diff
@@ -32,8 +32,26 @@ Refer to the [Generative AI Hub Supported Models page](https://me.sap.com/notes/
...
+- **Simplified usage:** provides access to all available models without requiring individual deployments.
```
Comment (ncryptedV1) [inline docs/provider-config/sap-aicore.mdx]:
```suggestion
- **Simplified usage:** provides access to all available models without requiring individual deployments using the [Harmonized API](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/harmonized-api)
```

File: docs/provider-config/sap-aicore.mdx [hunk 4]
```diff
@@ -32,8 +32,26 @@ Refer to the [Generative AI Hub Supported Models page](https://me.sap.com/notes/
...
+- 
+
```
Comment (ncryptedV1) [inline docs/provider-config/sap-aicore.mdx]:
```suggestion 

```

File: docs/provider-config/sap-aicore.mdx [hunk 5]
```diff
@@ -32,8 +32,26 @@ Refer to the [Generative AI Hub Supported Models page](https://me.sap.com/notes/
...
+**Native API Mode:**
+- **Available on all plans:** Works with any SAP AI Core service plan
+- **Manual deployments:** Requires manual deployment and management in ai core service instance
```
Comment (ncryptedV1) [inline docs/provider-config/sap-aicore.mdx]:
```suggestion
- **Manual deployments:** Requires manual model deployment and management in your SAP AI Core service instance
```

**Comment by lizzzcai** [comment]
why the orchestration mode only for extended plan, not for other plans?  I think it should provide to all and have option opt in or out. so existing one can either stay in the native api or switch to orchestration if they want.

File: docs/provider-config/sap-aicore.mdx [hunk 6]
```diff
@@ -32,8 +32,26 @@ Refer to the [Generative AI Hub Supported Models page](https://me.sap.com/notes/
...
+- In case your service plan is `extended`, keep Orchestration Mode enabled unless you have specific requirements that necessitate direct AI Core API access or need features not supported by the orchestration mode.
+
+- In case your Orchestration Mode is disabled:
+    -   **Model Selection:** The model dropdown displays models in two separate lists:
+        -   **Deployed Models:** These models are already deployed in your specified resource group and are ready to use immediately.
+        -   **Not Deployed Models:** These models don't have active deployments in your specified resource group. You won't be able to use these models until you create deployments for them in SAP AI Core.
+    -   **Creating Deployments:** To use a not deployed model, you'll need to create a deployment in your resource group in SAP AI Core service instance. See [Create a Deployment for a Generative AI Model](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/create-deployment-for-generative-ai-model-in-sap-ai-core) for instructions.
```
Comment (ncryptedV1) [inline docs/provider-config/sap-aicore.mdx]:
```suggestion
    -   **Creating Deployments:** To use a model that has not been deployed yet, you'll need to create a deployment in your SAP AI Core service instance. See [Create a Deployment for a Generative AI Model](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/create-deployment-for-generative-ai-model-in-sap-ai-core) for instructions.
```

File: src/api/providers/sapaicore.ts [hunk 7]
```diff
@@ -338,6 +344,7 @@ export class SapAiCoreHandler implements ApiHandler {
 	private options: SapAiCoreHandlerOptions
 	private token?: Token
 	private deployments?: Deployment[]
+	private aiCoreEnvSetup: boolean = false
```
Comment (ncryptedV1) [inline src/api/providers/sapaicore.ts]:
```suggestion
	private isAiCoreEnvSetup: boolean = false
```
nit + personal preference

**Comment by ncryptedV1** [comment]
> why the orchestration mode only for extended plan, not for other plans? I think it should provide to all and have option opt in or out. so existing one can either stay in the native api or switch to orchestration if they want.

@lizzzcai it's part of SAP's pricing plan, nothing we can affect. only the `extended` plan includes Generative AI Hub capabilities which contain the orchestration service with the harmonized API. for reference, see the [Service Plans documentation](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/service-plans)

File: src/api/providers/sapaicore.ts [hunk 8]
```diff
@@ -428,6 +435,94 @@ export class SapAiCoreHandler implements ApiHandler {
...
 
 	async *createMessage(systemPrompt: string, messages: Anthropic.Messages.MessageParam[]): ApiStream {
+		if (this.options.sapAiCoreUseOrchestrationMode ?? true) {
+			yield* this.createMessageWithOrchestration(systemPrompt, messages)
+		} else {
+			yield* this.createMessageWithDeployments(systemPrompt, messages)
+		}
+	}
+
+	// TODO: support credentials changes after initial setup
```
Comment (ncryptedV1) [inline src/api/providers/sapaicore.ts]: does this imply one has to switch providers away & back to SAP AI Core for this to re-run? if so, this will lead to confusion and bad UX.

File: src/api/providers/sapaicore.ts [hunk 9]
```diff
@@ -428,6 +435,94 @@ export class SapAiCoreHandler implements ApiHandler {
...
+		}
+
+		// Validate required credentials
+		if (
+			!this.options.sapAiCoreClientId ||
+			!this.options.sapAiCoreClientSecret ||
+			!this.options.sapAiCoreTokenUrl ||
+			!this.options.sapAiCoreBaseUrl
+		) {
+			throw new Error("Missing required SAP AI Core credentials for orchestration mode. Please check your configuration.")
```
Comment (ncryptedV1) [inline src/api/providers/sapaicore.ts]:
```suggestion
			throw new Error("Missing required SAP AI Core credentials. Please check your configuration.")
```
All are required regardless of orchestration mode. Solely the `AICORE_SERVICE_KEY` env var is implicitly used by `@sap-ai-sdk/orchestration` using `@sap-ai-sdk/ai-api` under the hood

File: src/core/controller/models/getSapAiCoreModels.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+		return { modelNames, orchestrationAvailable }
+	} catch (error) {
+		console.error("Error fetching deployments:", error)
+		throw new Error("Failed to fetch deployments")
+	}
+}
+
+/**
+ * Fetches available models from SAP AI Core deployments with orchestration availability
```
Comment (ncryptedV1) [inline src/core/controller/models/getSapAiCoreModels.ts]:
```suggestion
 * Fetches available models from SAP AI Core deployments and orchestration availability
```
nit

File: webview-ui/src/components/settings/providers/SapAiCoreProvider.tsx
Comment (ncryptedV1) [inline webview-ui/src/components/settings/providers/SapAiCoreProvider.tsx:1]: Regarding refresh on settings change: Can't we just call `fetchSapAiCoreModels` on blur of any of the credential fields? Ideally, with a check whether the content has actually changed.

**Review by ncryptedV1 — APPROVED** [review]
Mostly nits

**Comment by mauriciolauffer** [comment]
@yuvalman, just a comment on

> for other plans (such as standard and free), keep the current behavior (without the "Orchestration Mode" checkbox) that shows only the Model list (separated to deployed and not deployed):

That's incorrect. SAP AI Core `standard` and `free` plans don't give you access to Generative AI Hub. The `extended` plan is the only one that gives you access to LLMs.

File: src/api/providers/sapaicore.ts (hunk 9 above)
Comment (yuvalman) [reply src/api/providers/sapaicore.ts]: I agree, 

I changed the error, and made it a func that will be called also before the authentication step that used in createMessageWithDeployments flow

File: src/api/providers/sapaicore.ts (hunk 8 above)
Comment (yuvalman) [reply src/api/providers/sapaicore.ts]: No,
It means that if a user starts a new Task using the credentials of service instance X, and then switches to the credentials of another service instance Y,
it will start failing in orchestration mode because the variable `this.isAiCoreEnvSetup` will already be marked as true (because it will still use X’s credentials and try to communicate with the Y service instance).

Starting a new task will fix the issue immediately for the user.

I believe this is a rare edge case, which is why I added it as a TODO. My plan was to implement the fix right after this PR is merged.

If you think it’s better to avoid this edge case entirely, we have two options:

1. Remove the logic of saving the credentials in this session - This would mean that in every createMessage call, the environment variable `AICORE_SERVICE_KEY` would be set up again. The edge case would disappear, but the setup would happen each time createMessage will be called.
2. Implement the planned fix in this PR before merging it - The downside is that it could take more time to complete.


Please let me know what do you think.

**Comment by ncryptedV1** [comment]
@mauriciolauffer ~I think you misunderstood this sentence as this is exactly what he is saying.~

~> for other plans (such as standard and free), keep the current behavior (without the "Orchestration Mode" checkbox)~

~This means that for standard & free plans, there will be no orchestration mode checkbox as there is no orchestration mode available in the first place. "Keep the current behavior" means falling back to regular AI Core API-based inference.~

Edit: Got your point, thanks for the nudge! You're right that the LLMs provided by SAP in the `foundation-models` scenario are only available with the `extended` plan. In very rare instances, users might deploy a specific model themselves in their own scenario where this would still be useful (beyond the caveats mentioned in the docs about specific requirements necessitating the use of the native API). Docs have been updated.

File: src/api/providers/sapaicore.ts (hunk 8 above)
Comment (ncryptedV1) [reply src/api/providers/sapaicore.ts]: Ah understood. I think then it's fine as is.

File: webview-ui/src/components/settings/providers/SapAiCoreProvider.tsx
Comment (yuvalman) [reply webview-ui/src/components/settings/providers/SapAiCoreProvider.tsx:1]: If desired, I can easily switch this to only affect the SAP AI Core Provider component, but I think this can add value to other providers, too - I opened a [discussion](https://github.com/cline/cline/discussions/5562) regarding making it reusable for other providers.

**Comment by yuvalman** [comment]
> @yuvalman, just a comment on
> 
> > for other plans (such as standard and free), keep the current behavior (without the "Orchestration Mode" checkbox) that shows only the Model list (separated to deployed and not deployed):
> 
> That's incorrect. SAP AI Core `standard` and `free` plans don't give you access to Generative AI Hub. The `extended` plan is the only one that gives you access to LLMs.

@mauriciolauffer , after discussing with @ncryptedV1 , he mentioned that as you said, without the Generative AI Hub, there is no access to the LLMs (that are only available in the `foundation-models` scenario)

I've updated the docs accordingly.

Please let me know what do you think.

**Comment by yuvalman** [comment]
Hi @celestial-vault ,
After we merge #5315 that was a blocker to this feature, can we please proceed with merging this feature?
It will improve the UX for SAP AI Core provider users

**Comment by celestial-vault** [comment]
@yuvalman Yeah will review in a bit

File: src/core/api/providers/sapaicore.ts [hunk 11]
```diff
@@ -178,7 +180,11 @@ namespace Bedrock {
 				imageData = new Uint8Array(Buffer.from(base64Data, "base64"))
 			} else if (item.source.data && typeof item.source.data === "object") {
 				// Try to convert to Uint8Array
-				imageData = new Uint8Array(Buffer.from(item.source.data as Buffer | Uint8Array))
+				if (Buffer.isBuffer(item.source.data)) {
+					imageData = new Uint8Array(item.source.data)
```
Comment (lizzzcai) [inline src/core/api/providers/sapaicore.ts]: this seems like not working with claude 4, can you help to cross check with this [PR](https://github.com/cline/cline/pull/5735).

Comment (yuvalman) [reply src/core/api/providers/sapaicore.ts]: @lizzzcai , done and tested in this pr too, now images can be sent to claude 4

**Comment by celestial-vault** [comment]
@lizzzcai @yuvalman Are you ready for review now?

File: src/core/storage/CacheService.ts [hunk 12]
```diff
@@ -246,6 +246,7 @@ export class CacheService {
 			sapAiCoreBaseUrl,
 			sapAiCoreTokenUrl,
 			sapAiResourceGroup,
+			sapAiCoreUseOrchestrationMode,
```
Comment (celestial-vault) [inline src/core/storage/CacheService.ts]: Curious, has this new service layer been clear to you all? Was it easy or difficult to figure out?

Comment (yuvalman) [reply src/core/storage/CacheService.ts]: > Curious, has this new service layer been clear to you all? Was it easy or difficult to figure out?

The orchestration layer abstracts provider specifics and normalizes responses to an OpenAI-compatible shape, so we don’t juggle per-model response formats (more details about the orchestration layer and it's features can be found in the [docs](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/orchestration-8d022355037643cebf775cd3bf662cc5)).

In Cline, we’re using the Harmonized API feature, which lets us call a range of models through one interface without deploying or maintaining them individually. This means the UX is simpler and setup is lighter.

Impact:
• User side: smoother onboarding and day-to-day use of the SAP AI Core provider in Cline.
• Dev side: one consistent response shape = fewer code paths and easier maintenance.

So if we have the orchestration mode, why do we need to keep the native API mode? Because some features are not supported by Orchestration Mode. If someone wants to leverage those features (such as prompt caching supported by Claude models or reasoning effort supported by OpenAI models), they can disable Orchestration Mode and use the native API mode instead.

Full explanations of the new mode are documented in `docs/provider-config/sap-aicore.mdx`


From a UI perspective, I exposed this as an “Orchestration Mode” checkbox. It seemed like the clearest on/off indicator. Open to renaming or switching to another affordance if preferred.

Comment (celestial-vault) [reply src/core/storage/CacheService.ts]: @yuvalman I meant the CacheService.ts

Comment (celestial-vault) [reply src/core/storage/CacheService.ts]: But thanks for the info on the Orchestration layer

**Comment by celestial-vault** [comment]
Needs a main merge

File: src/core/storage/CacheService.ts (hunk 12 above)
Comment (yuvalman) [reply src/core/storage/CacheService.ts]: @celestial-vault ,
Thanks for clarifying - You meant the cacheService / stateManager layer. It was easy to figure out and use, and it gives a really great performance improvement!

I also ran a small perf check and saw clear benefits for users experience and developers.

Comment (celestial-vault) [reply src/core/storage/CacheService.ts]: Awesome

**Comment by celestial-vault** [comment]
Would you mind resolving all conversations next time? 🙏

**Comment by yuvalman** [comment]
> Would you mind resolving all conversations next time? 🙏

Sure :)
